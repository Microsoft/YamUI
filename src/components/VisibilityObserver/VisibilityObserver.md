### Examples

Basic usage:

```js { "props": { "data-example": "basic" } }
const { GutterSize, TextSize } = require('../Block');

const renderInView = () => {
  return (
    <div style={{ backgroundColor: '#adedba', transition: 'background-color 1s' }}>
      <Block padding={GutterSize.XXLARGE} textSize={TextSize.LARGE}>In view</Block>
    </div>
  );
};

const renderOutOfView = (inView) => {
  return (
    <div style={{ backgroundColor: '#f98985', transition: 'background-color 1s' }}>
      <Block padding={GutterSize.XXLARGE} textSize={TextSize.LARGE}>Out of view</Block>
    </div>
  );
};


<VisibilityObserver 
  renderInView={renderInView}
  renderOutOfView={renderOutOfView}
  onEnter={action('scrolled into view')}
  onLeave={action('scrolled out of view')}
/>
```
