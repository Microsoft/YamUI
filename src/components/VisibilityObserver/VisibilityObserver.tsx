/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import Observer from 'react-intersection-observer';
import { BaseComponentProps } from '../../util/BaseComponent/props';

export interface VisibilityObserverProps extends BaseComponentProps {
  /**
   * A callback which will be triggered when the component is scrolled into view.
   */
  onEnter?: () => void;

  /**
   * A callback which will be triggered when the component is scrolled out of view.
   */
  onLeave?: () => void;

  /**
   * Render prop to return child content when the component is visible in the viewport.
   */
  renderInView?: () => React.ReactNode;

  /**
   * Render prop to return child content when the component is not visible in the viewport.
   */
  renderOutOfView?: () => React.ReactNode;

  /**
   * Wrapper element tag name.
   * @default 'div'
   */
  tag?: string;

  /**
   * A CSS margin string which pushes the intersection boundary further in or out of the viewport.
   * A positive value will expand the viewport threshold so your component is considered "in view" while it's
   * still offscreen, for example you could set thresholdOffset="200px" you want an image to start loading before
   * it scrolls into view. A negative value will move the boundary into the viewport, triggering "in view" after
   * it's already becoming visible.
   * "200px", "200px 0 50px 0", "-50px" are all valid values.
   */
  rootMargin?: string;
}

/**
 * IntersectionObserver allows for conditional child rendering and callbacks when scrolled in or out of view.
 */
export default class VisibilityObserver extends React.Component<VisibilityObserverProps> {
  public render() {
    const { rootMargin, tag } = this.props;

    return (
      <Observer
        tag={tag || 'div'}
        rootMargin={rootMargin}
        onChange={this.onVisibilityChange}
        render={this.getObserverChildren}
      />
    );
  }

  private getObserverChildren = (isVisible: boolean) => {
    const { renderInView, renderOutOfView } = this.props;
    if (isVisible && renderInView) {
      return renderInView();
    }

    if (!isVisible && renderOutOfView) {
      return renderOutOfView();
    }

    return null;
  };

  private onVisibilityChange = (isVisible: boolean) => {
    if (isVisible) {
      this.props.onEnter && this.props.onEnter();
    } else {
      this.props.onLeave && this.props.onLeave();
    }
  };
}
