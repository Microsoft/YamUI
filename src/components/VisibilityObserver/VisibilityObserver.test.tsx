/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import Observer from 'react-intersection-observer';
import { shallow, ShallowWrapper } from 'enzyme';
import VisibilityObserver, { VisibilityObserverProps } from './index';

type CallbackProp = ((inView: boolean) => void);
type RenderProp = ((inView: boolean) => React.ReactNode);

describe('<VisibilityObserver />', () => {
  let component: ShallowWrapper<VisibilityObserverProps, {}>;
  const renderInView = () => <span>IN VIEW</span>;
  const renderOutOfView = () => <span>OUT OF VIEW</span>;
  let onEnterCallback: jest.Mock<{}>;
  let onLeaveCallback: jest.Mock<{}>;
  let renderProp: RenderProp;
  let callbackProp: CallbackProp;

  beforeEach(() => {
    onEnterCallback = jest.fn();
    onLeaveCallback = jest.fn();
  });

  describe('with minimal options', () => {
    beforeEach(() => {
      component = shallow(<VisibilityObserver />);
      renderProp = component.find(Observer).prop('render') as RenderProp;
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    it('renders empty content via its in-view render prop', () => {
      expect(renderProp(true)).toMatchSnapshot();
    });
  });

  describe('with render, callback, rootMargin and tag options', () => {
    beforeEach(() => {
      component = shallow(
        <VisibilityObserver
          renderInView={renderInView}
          renderOutOfView={renderOutOfView}
          onEnterView={onEnterCallback}
          onLeaveView={onLeaveCallback}
          rootMargin="100px"
          tag="span"
        />,
      );
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    describe('Observer rendering', () => {
      let renderProp: RenderProp;

      beforeEach(() => {
        renderProp = component.find(Observer).prop('render') as RenderProp;
      });

      it('in view render matches its snapshot', () => {
        expect(renderProp(true)).toMatchSnapshot();
      });

      it('out of view render matches its snapshot', () => {
        expect(renderProp(false)).toMatchSnapshot();
      });
    });

    describe('Observer callbacks', () => {
      beforeEach(() => {
        callbackProp = component.find(Observer).prop('onChange') as CallbackProp;
        renderProp = component.find(Observer).prop('render') as RenderProp;
      });

      describe('when in view', () => {
        beforeEach(() => {
          callbackProp(true);
        });

        it('in view callback is properly called', () => {
          expect(onEnterCallback).toHaveBeenCalled();
        });

        describe('and scrolled back out of view', () => {
          beforeEach(() => {
            callbackProp(false);
          });

          it('the in-view render prop remains rendered', () => {
            expect(renderProp(true)).toMatchSnapshot();
          });
        });
      });

      describe('when out of view', () => {
        beforeEach(() => {
          callbackProp(false);
        });

        it('out of view callback is properly called', () => {
          expect(onLeaveCallback).toHaveBeenCalled();
        });
      });
    });
  });
});
