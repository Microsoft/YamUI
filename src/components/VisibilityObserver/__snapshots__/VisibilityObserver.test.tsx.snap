// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VisibilityObserver /> with all options Observer rendering in view render matches its snapshot 1`] = `
<span>
  IN VIEW
</span>
`;

exports[`<VisibilityObserver /> with all options Observer rendering out of view render matches its snapshot 1`] = `
<span>
  OUT OF VIEW
</span>
`;

exports[`<VisibilityObserver /> with all options matches its snapshot 1`] = `
<Observer
  onChange={[Function]}
  render={[Function]}
  rootMargin="100px"
  tag="span"
  threshold={0}
  triggerOnce={false}
/>
`;

exports[`<VisibilityObserver /> with minimal options matches its snapshot 1`] = `
<Observer
  onChange={[Function]}
  render={[Function]}
  tag="div"
  threshold={0}
  triggerOnce={false}
/>
`;

exports[`<VisibilityObserver /> with minimal options renders empty content 1`] = `null`;
